drivers/SerialBase.cpp

```diff
@@ -55,15 +55,17 @@ void SerialBase::format(int bits, Parity parity, int stop_bits) {
 
 int SerialBase::readable() {
     lock();
-    int ret = serial_readable(&_serial);
+    int ret = 0;
+    ret = serial_readable(&_serial);
```

Comment:
> 1. SerialBase.cpp: the change of separating initialization of 'ret' is to avoid the compiler optimizing it. It is observed that on ublox EVK NINA-B1, 'ret' is always returning false as it has been optimized.

I fail to see how this is different - how this can produce a different output by a compiler. Are there any optimization that you are seeing with debug/release here? I recall from your report that return value is set to false in the release build.

The output should be the same as in the current form. Isn't the problem in the below layer - HAL ? How `serial_readable` is implemented for your target that you are using? @

Reply:
the serial_readable() on this platform actually references a certain register, which may look 'constant' to the compiler in O3 optimization profile in my experience. 
I've tried three ways: (all with release profile)
1. original -> serial_readable() always returns false
2. separating initialization (ret = 0; ret = serial_readable(), this is trying to tell the compiler that the value of this variable may change) -> works
3. adding volatile to ret in the original code -> works
I didn't disassembly the binary to see the actual assembly code, but the three binaries actually differs: the (2) and (3) are pretty close, but (2) and (3) are far different from (1)

Reply:
hmm, I'm using GCC ARM 6-2017-q2 update, and I don't see other messages running. I doubted the same thing before I get the ublox board as you commented that there are a bunch of volatile fields in the low level implementation which looked okay, but when I get the board and gave it a try, adding volatile or touching the ret variable in the caller actually helps.

Reply:
If this really were to be a compiler bug, which I don't think we've concluded yet, I would want to see it fixed by a pragma lowering optimisation in `serial_readable` or maybe `nrf_uart_event_check` - or maybe even a global toolchain option change.

This sort of "spelling change that really shouldn't cause a difference" in the caller doesn't really cut it as a workaround - it's too inexplicable why it should work, and could easily be accidentally "unfixed" later.

Reply:
as syntax of the pragma to change the optimization level varies among compilers, is there a suggested way to do that?
btw, confirming if this is actually a compiler bug requires a look at the assembly code, is there a way to do that without hardware debugger?

Reply:
@soleilplanet `-O3` is not used in any of our profiles. 

I've isolated the readable member function and the [code generated](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAKxAEZSBnVAV2OUxAHIB6bgagAqATwAOmPkJbE%2BaLHwSZimUnxJ8ANqgCG6PloB2fTAA8tAWxHrMAOgCkABgCCD5wCYAzHn3J1zObfcAYQYCdCwAM2sEAOwXWw8vHz9xAODQrwIomLineM9wiL4AfSLkS2YGcoYXPj48iK9xWpKASVr2gDdUdS0CPCt2waG%2BbgAqMDBUuoBWQIAjYjxMcL4AEWXGhj4wTiUdVX11IR2%2BMWIzPAYGPFR9LdHuOI9MdQZMGrqPBv0m4qK2wZdHp9AZoO4EYYjcaTIIzeaLZZrDY/LY7Pa6W5HE5nC5XG53PgPJ7uTD6fDhYnfX7NIoAeSGQN6/WpkNqYwmU1sswWSxW63Cm22nAA7osCOJMcdOKdFLjrrd7o9cl9kSzWvTOt0mQNWYN2TDAnCeYj%2BYK0Zh9vxRYRMNjZZd5QSibknGM%2BOFupphV5gHwqVsGAgWOpdHNxBVMLoPdIQsRmMgCKxxGZMGYw8RFZTVUNWgBZBlakHiMEhYb6znchF81Wo3YWjGHKV8WPxxNKPgptOKGXnB34zPK9xUnN0/OAwvM3Xtcuwrnw3lIgUooXW8UHLHSlsJpMd1Ppnty/uEpVuIfZwbqse1RlFqds6EV%2BcmmtC9EjPir22bghx7ftzv7jifYKseOSnngKxUroRS5gAyoERQAGrYAASp8ABs8TodsYBQcUgQoQIRQAKqwdgRQAELES0AAyAgtAAcnSABizFkQItLMR8tR5BBhQlGUviVEJ3FDPU2aoOE4RvEQ4QQAwpBmAAlHwUDXAAXpgRQEEpWFKBkigiEoBClFoISpMgCBaNIJYQjezJYTEUBqaMSn2EpAC0MTKUpulOIMeQvG8okBSqS4SlJMmSfJikqWpeCadpumuOhzkMK57lee42A%2BaJgVkhBxJBe8/ntOJ4WqJFmCyTFyl/HMzD9H0%2BhFJJ0nVdFCnKUVBUUi6ziDsODEocxJGOIR9iUY4ZF3pC9jGAALPYy2uMt9jETRYEEKIkaIluEK2AA7BRs2nXeM6GsR40CHwsE/q2O5nU9z1Ts6jg0hqjX6AQ7iuNptQCNNADSsFFLBgOEShAAaAQnS904PrCd3WRCV2EXwSjsHgHTdvDeOnW9H21F9P1/RCgjA6D4O0gACtDsP4/eHJI0QIh8GjN2Y5g2O44zfOQoTfyfRkv3/RTsEg2DEMCAIMPuHDeMXbdBAo%2Bz118D%2BBgMBcBDitI/MG4LJTC99ovk4DEtUxxNOywzjNK3dqBsxzGvEFrOt6wbXtG/8xMi2TtQodgZEochqz2FyFHuFyqywx8xt%2B6bAfi5LsGkTT2AMbH8v24jhqwRUYhkmr6Ne2Xww%2BwCJNm4HweoWHtCR9MmHTNnFHx/89LV8n2DIQxAig4EA923zStD7BfCXHoCbY70kaqTJGioMKSnWHwgRWNZgioLdpLoGv5eV13/ti73mcD0UDHjyP%2BNjwPk9bFgWgzx0c%2B6PJ1XyHgwAIKvfAMagCEG8LTSAEDvWCe8D54yPnwbup8%2B4X2hqsFCqwACaN8XpK1WL0LQGNMBYxxroLwfAkHl35pXROpMxZBxDmHVwkcFoxzjqVdUsCT7kzPv3UGstkFoIwc9LBODmykghOEYgqAzCCChqsMhhsTw0iruw2utDsCrGjvLDuLRj5J3gefUGqEUK0hQvwgRec%2BDYGIOI6QWBxQJnnrIsuFC2E6PJjQ%2BuqjGHTAoodJhGiWGd2cVQjhCDQbQw4iY0xzNDQoXwdzHG0g%2BgphYOTBxfMnFwNcXXUOqjpgMJbm3TR2igm1FggACSMcPHO5cHZBmIAQZAzAIRKGAJcT2qTGbpKUSQrJYcW4UXQp4gp/itGBJrnwRiAhM7sQifDJW2B9BaDmAMAyljmAiBSe06B8i/gjIybUCZmdAg0WMVUg2WDLiLOWd9RQcZ1mbPxp0lxyj3GrB8V4gAnNHVuzD3o7KKWMgxRjYIoUCDMzBZiLFWObFIdg9y5GaMoWMtx2TVgAA5I7uEbt8vxvzWF7PMQxRwFEaKxFOV7OZCylniBdnC8h2zWiIuTsisO7yfkKP%2BcnGmZFjmVIVmcsxNNiFvCsAmd0ahCIT1pU9GB%2BKuXYDotIsF4Kol8EFYYYV%2BDRFqB4VK6V9KAmyu5dfMl/MlZquESKrV0hx66rOjKrpcrjmKpNaPAVQqXiarFdIUhtrZqPOKSQ51fLqlmNIc01pvNfV6o7ibANPClW3zMTq8NIRI1RoJvqxRTzukqNWLQCOOL2WjOThRRwxFkGOEmQms6SsKJaD8BjOe6atkIuLdQnpqjaBYoorQN5QzcUGq6YEWkDFmItAAOLVqemPW4ApgCsCZLcSqpxrKECEHoYuVliDoGFNZcQ4QvQyFuD%2BboHSTxHRkcNUaHMijCDED8lw%2B1hGLC0OobSD7BrZivWNCa/MoDfpvXelkFcVIAeupNUtZE/KnmHDexi4NHBX2wJCMDE0to7UKKSZgZg4jHVEmMdQGQ%2BAeQYFoHGxHMDuHsILWoqGiKcKIjasq7gZFtSinJOjt6doqAY4PAeSkVD3lGFMbAONvrunEZIm1hGfjTz6K/Potw140f/iNH99GEGXyY3UFjlV2o1U40BnjIStP8cE2MYTsJRMiIkxI9e98ZPiCfi/Rd%2BhlPbNo2pm9DGihIJQagnTrGqoGa89dLjYhjN6N89I/zAmLMibE6IyTfohHonXcQpB7nRKcZ8zw/zgW9PsYgIZ7j5iTN5bQXF0YlnDTWfE2Iuz6AhFvHq8lnhWXSqeevWFnzgK0IFbYx1DjoXCLheUGVqLfWBNQhq%2BYxLtnJE3LUDYzV%2BIOu/K6%2BpooPmwkagG8F6KJWIsTa4dFjiKkhMJZsw1yRXMeYJLwEkxp629SjBkxCDySgQgkEwG9C9fB9DEHCEUZgKMiiYES%2B%2BnFYEwJzFQKewHwPQd1PB5Dyy%2BCADWxWRtESA4SU4RRmkqERyDsHEORH/XJ99aDR126dbwW2QwEA4fdCUqMCA9kBh7NclAEmqL/quREITzAvp4gnQgBkpSVOdIPsOrHfqLgiNvBfW%2B9ElzMAfxVwL1AcwqA07w04DnqA8DoCUjrqgD6Eh8l7i0QI5EQ4tEcDRIo01UFX1KaJCCqkb2BAotYYgxgijP3k5gFStOQrtGMqwQwBa6cbbqHLnq5JRJR%2BIEzknyOTLS9KAoZAWO4MMQQ0hlQOXNN%2Bcq7paHif%2BrK7wK%2B7Soxii19fR%2BxwXQTdLzzxAMPx0E/y7b8b3QzBDioC7z3k6/2wKT3E2rql3eE9x8GJoMfbL2hEeMgV5vqv6zq4gEULflfF/tGH8vrHh%2BU/VWjwztlF6wKcAE%2BoLg0xOCkH0FwewL/UBcFBa4duv/oWsCwp5C0Av4EDv734CYY4gDTD2CkCP6cALQv5v6cAf6kBf6cAv4MAgCwFgEoH36kBwCwBIBoAWDMhkAUAQAkEiBkEgDWRmDADoRUakACjqB6xYF74lD2CUACZzDgGkBzBeDWRCBcAgGkAkEpjfS0iNh8FYBmAGDABWAyF4CYwh5YF4GkAmD4KNIcCcCiEGTwGoFUAACOzAigwhnA7g1grg1gohVg%2BgwABASA9ARkEOtIuumqlAJQRQXB8kCU6kVgtAMB0w7ytA7yrgC0qK7gC0rglAw%2BUElAVgwAe88A0wpAsBuhEBvkcBpIDhThpALhHQbhVAHhHB3hlAVAAAiqYcQEIL9Pmr2ktKtF2sEa4LQKisEdMLQJQPgMJFoOYfgRpP4ZgIEfYMEaEeEZEdEbEWSKqOgAkSLskZAKkekSAb5FkXYbkXQFkT%2BI9iIfgYRnMFgZAAJk7IpncFwB5LSO4MRiEOgAEKsMgGLmLu8qinwJgTChwLQA/k/kgXwegeNPmMAMgMgHwOhNYJYfYKpDJswMYHFLgIQGoMASoIEBItQVYNIMAUpKAeAVkVATAXAVwIga/n8VwJgdgaQLgR/usVwK4L8eoegdiXgVkfEo6CAAtEAA) with GCC 6.3 seems fine.

> 1. original -> serial_readable() always returns false
> 2. separating initialization (ret = 0; ret = serial_readable(), this is trying to tell the compiler that the value of this variable may change) -> works
> 3. adding volatile to ret in the original code -> works

(1) and (2) should not be different; ret and is address is not used while 0 is assigned to it therefore, the 0 assignment can be skipped. Of course (3) is different.

> btw, confirming if this is actually a compiler bug requires a look at the assembly code, is there a way to do that without hardware debugger? 

You can use Objdump to dump the code of the function.

Reply:
Looks like O3 is used for ARM and uARM in develop and release profiles.
* https://github.com/ARMmbed/mbed-os/blob/master/tools/profiles/release.json#L27-L45
* https://github.com/ARMmbed/mbed-os/blob/master/tools/profiles/develop.json#L27-L45

Reply:
I forgot to precise gcc profiles since it is what @soleilplanet use apparently. 

@soleilplanet Have you test with other compilers and reproduced the issue ?

Reply:
@pan- I don't see this problem with GCC ARM 5.4.1 on ubuntu either, this happens on Windows environment.
For the disassembly, I just realized that there is a  symbol map file generated, I'll check that with objdump.